### PR TITLE
fix(controller): set pod finish timestamp when it is deleted

### LINKED ISSUE #2901: Retry Strategy doesn't work in GKE preemptible node (pod deleted)
Checklist:

 * [x] I've included the version.
 * [ ] I've included reproduction steps.
 * [ ] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:
I configured the retry strategy as the following
```
retryStrategy:
      limit: 4
      retryPolicy: "Always"
      backoff:
        duration: "12s"
        factor: 5
        maxDuration: "200m"
```
and run the Argo workflow in a GKE preemptible nodes.
When GKE required resources back deleting the nodes, the retry strategy doesn't work.
I received a red warning /!\ without the retry.

**What you expected to happen**:
I expected the retry strategy to works properly because I setup **retryPolicy: "Always"** as well.

**How to reproduce it (as minimally and precisely as possible)**:
Create a workflow with the following retry Strategy and run it in GKE GCP in a preemptible node
```
retryStrategy:
      limit: 4
      retryPolicy: "Always"
      backoff:
        duration: "12s"
        factor: 5
        maxDuration: "200m"
```

**Anything else we need to know?**:

**Environment**:
GKE (GCP)
- Argo version: v2.7.6
```
$ argo version
argo: v2.7.6+70facdb.dirty
  BuildDate: 2020-04-28T17:09:06Z
  GitCommit: 70facdb67207dbe115a9029e365f8e974e6156bc
  GitTreeState: dirty
  GitTag: v2.7.6
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2020-02-06T03:31:35Z"
  compiler: gc
  gitCommit: f5757a1dee5a89cc5e29cd7159076648bf21a02b
  gitTreeState: clean
  gitVersion: v1.14.10-dispatcher
  goVersion: go1.12.12b4
  major: "1"
  minor: 14+
  platform: darwin/amd64
serverVersion:
  buildDate: "2020-02-21T18:01:40Z"
  compiler: gc
  gitCommit: 145f9e21a4515947d6fb10819e5a336aff1b6959
  gitTreeState: clean
  gitVersion: v1.14.10-gke.27
  goVersion: go1.12.12b4
  major: "1"
  minor: 14+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:

- workflow result:
Screen Shot 2020-04-30 at 1 59 01 pm

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by jiezhang**
Summary
----
Fixes #2901

When a node fails due to a pod-deleted error (see code [here](https://github.com/argoproj/argo/blob/a7d8546cf9515ea70d686b8c669bf0a1d9b7538d/workflow/controller/operator.go#L833)), `FinishedAt` is unavailable and retry backoff doesn't work as expected. Backoff is crucial to handle the pod-deleted error since it usually happens when the cluster is under pressure.

This PR sets `FinishedAt` to system time when the pod is deleted. 

Tests
----
Verified with our internal load test:
* 3600 workflows
* 2 workflows created per second

Before: finishedAt is null, and workflow success rate is 66.1%.
```
foo-pipeline-xhwq8-3799176537:
  boundaryID: foo-pipeline-xhwq8-2284472225
  displayName: echo(2)
  finishedAt: null
  id: foo-pipeline-xhwq8-3799176537
  message: pod deleted
  name: foo-pipeline-xhwq8.exit-handler-1.echo(2)
  phase: Error
  startedAt: "2020-06-14T18:07:54Z"
  templateName: echo
  type: Pod
```

After: finishedAt is available, and workflow success rate is 95.7%.
```
foo-pipeline-nltcq-2896542097:
  boundaryID: foo-pipeline-nltcq-1123147699
  displayName: sleep(0)
  finishedAt: "2020-06-18T18:29:11Z"
  id: foo-pipeline-nltcq-2896542097
  message: pod deleted
  name: foo-pipeline-nltcq.exit-handler-1.sleep(0)
  phase: Error
  startedAt: "2020-06-18T18:23:37Z"
  templateName: sleep
  type: Pod
```

Checklist
----
* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

**Comment by jiezhang** [comment]
@alexec @sarabala1979 @simster7 Can you please take a look at this PR?

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -678,7 +678,11 @@ func (woc *wfOperationCtx) processNodeRetries(node *wfv1.NodeStatus, retryStrate
 			// Formula: timeToWait = duration * factor^retry_number
 			timeToWait = baseDuration * time.Duration(math.Pow(float64(retryStrategy.Backoff.Factor), float64(len(node.Children))))
 		}
-		waitingDeadline := lastChildNode.FinishedAt.Add(timeToWait)
```
Comment (alexec) [inline workflow/controller/operator.go:681]: @jessesuen I'd love your eyes on this please.

Comment (jiezhang) [reply workflow/controller/operator.go:681]: `MaxDuration` currently counts from `StartedAt`: https://github.com/argoproj/argo/pull/2976/files

Comment (jiezhang) [reply workflow/controller/operator.go:681]: We did a stress test recently. Without proper backoff, up to 30% of the workflows failed due to the pod-deleted error.

**Comment by simster7** [comment]
Hi @jiezhang, this was already fixed in https://github.com/argoproj/argo/pull/2976. Closing as fixed.

**Comment by simster7** [comment]
Sorry, I just realized that I misinterpreted this PR

**Comment by jiezhang** [comment]
If we introduce a new field, this bug will persist until users explicitly specify the new field. Given that the issue might be difficult to understand and does not reproduce under normal circumstances, I am concerned that most users will omit the new field and continue to experience this issue. 

Alternatively, we could use FinishedAt by default and fall back to StartedAt if that’s unavailable. The problem with that approach is that the backoff duration will be likely too short for the backoff to kick in. We might need to for example double the duration to make room for the backoff, which makes the behavior even harder for end users to understand. 

Counting from StartedAt is definitely not a perfect solution, but its definition will be consistent with MaxDuration. Since this field can be adjusted on every step and users generally have a rough idea of the execution time, users should be able to come up with a working number.

**Comment by jiezhang** [comment]
Maybe we could set FinishedAt to system time when the node is marked deleted?

https://github.com/jiezhang/argo/blob/e4343cc4a8ca9153d3f952b7fdd806e8d70d3db6/workflow/controller/operator.go#L837

**Comment by simster7** [comment]
Let me mull it over and discuss it with the team tomorrow and I'll report back. We definitely want this fixed ASAP as other users are having similar issues.

cc @fabio-rigato FYI: I am holding off on the 2.8.2 release until we can solve this issue which is similar to #2976

**Comment by fabio-rigato** [comment]
thanks so much, @simster7 for your effort in trying to fix these issues!
yeah maybe all of them are related!
all the best
Fabio

**Comment by simster7** [comment]
> Maybe we could set FinishedAt to system time when the node is marked deleted?

@jiezhang This seems like the best solution. Could you update the PR please?

**Comment by simster7** [comment]
@jiezhang, let me know if you are able to do this today, I'd want to get this merged and 2.8.2 released soon. If not, I can take this up without a problem

**Comment by jiezhang** [comment]
> @jiezhang, let me know if you are able to do this today, I'd want to get this merged and 2.8.2 released soon. If not, I can take this up without a problem

The code change is straightforward, but I still need to run the stress test again to verify it.

**Comment by simster7** [comment]
> The code change is straightforward, but I still need to run the stress test again to verify it.

Sounds good. Let me know your results

**Comment by jiezhang** [comment]
@simster7 @alexec @fabio-rigato I have verified the latest fix and updated the PR description.

**Comment by fabio-rigato** [comment]
Thanks, @jiezhang and @simster7 for your effort in that, really appreciated!
I will try the new release as soon as possible and back to you if I will find any other similar issues.
Hope you guys a great weekend!

**Comment by simster7** [comment]
@jiezhang @fabio-rigato v.2.8.2 is released with this fix and #2976: https://github.com/argoproj/argo/releases/tag/v2.8.2

**Pull request merged**